iroh/examples/sync.rs

```diff
@@ -0,0 +1,1057 @@
+use url::Url;
+
+use iroh_bytes_handlers::IrohBytesHandlers;
+
+const MAX_DISPLAY_CONTENT_LEN: u64 = 1024 * 1024;
+
+type Doc = Replica<<store::fs::Store as store::Store>::Instance>;
+
+#[derive(Parser, Debug)]
+struct Args {
+    /// Private key to derive our peer id from
+    #[clap(long)]
+    private_key: Option<String>,
+    /// Path to a data directory where blobs will be persisted
+    #[clap(short, long)]
+    storage_path: Option<PathBuf>,
+    /// Set a custom DERP server. By default, the DERP server hosted by n0 will be used.
+    #[clap(short, long)]
+    derp: Option<Url>,
+    /// Disable DERP completeley
+    #[clap(long)]
+    no_derp: bool,
+    /// Set your nickname
+    #[clap(short, long)]
+    name: Option<String>,
+    /// Set the bind port for our socket. By default, a random port will be used.
+    #[clap(short, long, default_value = "0")]
+    bind_port: u16,
+    /// Bind address on which to serve Prometheus metrics
+    #[clap(long)]
+    metrics_addr: Option<SocketAddr>,
+    #[clap(subcommand)]
+    command: Command,
+}
+
+#[derive(Parser, Debug)]
+enum Command {
+    Open { doc_name: String },
+    Join { ticket: String },
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let args = Args::parse();
+    run(args).await
+}
+
+pub fn init_metrics_collection(
+    metrics_addr: Option<SocketAddr>,
+) -> Option<tokio::task::JoinHandle<()>> {
+    iroh::metrics::init_metrics_collection();
+    // doesn't start the server if the address is None
+    if let Some(metrics_addr) = metrics_addr {
+        return Some(tokio::spawn(async move {
+            if let Err(e) = iroh_metrics::metrics::start_metrics_server(metrics_addr).await {
+                eprintln!("Failed to start metrics server: {e}");
+            }
+        }));
+    }
+    tracing::info!("Metrics server not started, no address provided");
+    None
+}
+
+async fn run(args: Args) -> anyhow::Result<()> {
+    // setup logging
+    let log_filter = init_logging();
+
+    let metrics_fut = init_metrics_collection(args.metrics_addr);
+
+    // parse or generate our keypair
+    let keypair = match args.private_key {
+        None => Keypair::generate(),
+        Some(key) => parse_keypair(&key)?,
+    };
+    println!("> our private key: {}", fmt_secret(&keypair));
+
+    // configure our derp map
+    let derp_map = match (args.no_derp, args.derp) {
+        (false, None) => Some(default_derp_map()),
+        (false, Some(url)) => Some(DerpMap::from_url(url, 0)),
+        (true, None) => None,
+        (true, Some(_)) => bail!("You cannot set --no-derp and --derp at the same time"),
+    };
+    println!("> using DERP servers: {}", fmt_derp_map(&derp_map));
+
+    // build our magic endpoint and the gossip protocol
+    let (endpoint, gossip, initial_endpoints) = {
+        // init a cell that will hold our gossip handle to be used in endpoint callbacks
+        let gossip_cell: OnceCell<Gossip> = OnceCell::new();
```

Comment:
not that this needs to change, but i believe the "async way" is to use a oneshot channel for this.